linux_os/guide/system/network/network-firewalld/ruleset_modifications/ensure_firewall_rules_for_open_ports/oval/shared.xml

```diff
@@ -1,76 +1,183 @@
+    <linux:protocol operation="pattern match">^.*$</linux:protocol>
+    <linux:local_address operation="equals">::1</linux:local_address>
+    <linux:local_port datatype="int" operation="greater than or equal">0</linux:local_port>
+  </linux:inetlisteningservers_state>
+
+  <linux:inetlisteningservers_state id="state_inet_foreign_port_connected" version="1"
+    comment="Checks this is a listening service not connected">
+    <linux:foreign_port datatype="int" operation="not equal">0</linux:foreign_port>
+  </linux:inetlisteningservers_state>
+
+  <!-- Objects and variables describing the firewalld configuration -->
+  <ind:textfilecontent54_object id="object_active_firewalld_zone_cfgs" version="1">
+    <ind:path>/etc/firewalld/zones/</ind:path>
+    <ind:filename operation="pattern match">^.+\.xml$</ind:filename>
+    <ind:pattern operation="pattern match">^[\s]*&lt;interface\s*name=.*$</ind:pattern>
+    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <local_variable id="var_firewalld_active_zones" version="1"
+    comment="Firewalled ports according to firewalld configuration" datatype="string">
+    <object_component object_ref="object_active_firewalld_zone_cfgs" item_field="filename"/>
   </local_variable>
 
+  <ind:textfilecontent54_state id="state_firewalld_zones_with_interfaces" version="1"
+    comment="Consider only active zones (i.e. with interfaces assigned)" >
+    <ind:filename operation="equals" var_ref="var_firewalld_active_zones" var_check="at least one"/>
+  </ind:textfilecontent54_state>
 
-  <local_variable id="var_firewalled_services" version="1" comment="Firewalld service names" datatype="string">
+  <ind:textfilecontent54_object id="object_firewalled_service" version="1">
+    <ind:path>/etc/firewalld/zones/</ind:path>
+    <ind:filename operation="pattern match">^.+\.xml$</ind:filename>
+    <ind:pattern operation="pattern match">^[\s]*&lt;service name="(\S*)".*$</ind:pattern>
+    <ind:instance operation="greater than or equal" datatype="int">1</ind:instance>
+    <filter action="include">state_firewalld_zones_with_interfaces</filter>
+  </ind:textfilecontent54_object>
+
+  <local_variable id="var_firewalled_services" version="1"
+    comment="Firewalld service names" datatype="string">
     <object_component object_ref="object_firewalled_service" item_field="subexpression"/>
   </local_variable>
 
-  <local_variable id="var_all_firewalled_services_desc_filename" version="1" comment="Firewalld service file description" datatype="string">
+  <local_variable id="var_all_firewalled_services_desc_filename" version="1"
+    comment="Firewalld service file description" datatype="string">
     <concat>
       <variable_component var_ref="var_firewalled_services"/>
       <literal_component>.xml</literal_component>
     </concat>
   </local_variable>
 
+{{% set inet_protos = ['tcp', 'udp'] %}}
```

Comment:
nitpick: Maybe incorporate use the protocols directly in the loop below? I think `tcp` and `udp` are pretty static values for this loop, so it is not expected this list to become larger in the future.

Reply:
Not sure I understand what do you mean by `incorporate use the protocols directly in the loop below?` the protocols list is static also in my opinion, but not sure what other approach to use with jinja, besides defining a list in the file?

Reply:
`{{% for proto in ['tcp', 'udp'] %}}`

Reply:
:) that totally blew :exploding_head:  my 70s C programmer brain. `Declare variables in the beginning. Don't mess up with the stack! ` :)
To put the the stupid jokes aside, actually you are right but now if I am using same up in the criteria section. Maybe still better having the `{{% set inet_protos = ['tcp', 'udp'] %}}` list in the beginning of the file and use it both for criteria and objects/variables/test/definition. I know it static but feels wrong to have code duplication as result of refactoring :)